Update notes for AWS object storage read/write (#183)
hytest-org/hytest: essential_reading/DataSources/Data_S3.md

# Data / Cloud Storage

One of the main storage locations for HyTest data is in '_The Cloud_'. This is sometimes referred to as **'Object Storage'**.

> The data is kept in datacenters operated by Amazon, Microsoft, or similar, which makes it easily available to network-connected devices.

Reviewer: seems two words is the more common and accepted spelling
```suggestion
The data is kept in data centers operated by Amazon, Microsoft, or similar, which makes it easily available to network-connected devices.
```

> The main advantage of doing this is that if your compute engine is also in that same datacenter (or nearby, as is the case with many JupyterHub services),

Reviewer: ```suggestion
The main advantage of doing this is that if your compute engine is also in that same data center (or nearby, as is the case with many JupyterHub services),
```

the data doesn't have to go very far to get to the compute power.
This brings the computation to the data, rather than shipping large datasets across the internet to get to the compute engine.

> [S3](https://aws.amazon.com/s3/) is Amazon's implementation of object storage, which pairs with the Amazon (AWS) nodes on which the Jupter Hub runs.

Reviewer: ```suggestion
[S3](https://aws.amazon.com/s3/) is Amazon's implementation of object storage, which pairs with the Amazon (AWS) nodes on which the JupyterHub runs.
```

> [S3](https://aws.amazon.com/s3/) is Amazon's implementation of object storage, which pairs with the Amazon (AWS) nodes on which the JupyterHub runs.
> What follows is a brief demo of how S3 data is accessed (both read and write), and some pitfalls to watch out for.

Reviewer: ```suggestion
The HyTEST project purchased a 1 PB [Open Storage Network (OSN)](https://www.openstoragenetwork.org/) pod, providing [Ceph Object Storage](https://docs.ceph.com/en/pacific/glossary/#term-Ceph-Object-Storage) on a high-speed network. Because we own this piece of hardware, we can host data without the storage or egress fees that come along with other forms of cloud object storage, such as [Amazon S3 object storage](https://aws.amazon.com/s3/).

Ceph object storage supports an [API that is compatible with the basic data access model of the Amazon S3 API](https://docs.ceph.com/en/pacific/radosgw/s3/#).

What follows is a brief demo of how data can be read from or written to the OSN pod, and some pitfalls to watch out for.
```

Reviewer: @rsignell-usgs - would you like you to take a glance at what I drafted here

The permissions scheme for S3 allows for anonymous/global read access, as well as secured access via specific credentials.
We'll look at generic workflows using an anonymous-access store, then finish off with some private/credentialed operations.

> The easiest way to access S3 data from within a Python program is via the

Reviewer: ```suggestion
The easiest way to access data on the OSN pod from within a Python program is via the
```

[fsspec](https://filesystem-spec.readthedocs.io/en/latest/) module.
This is a layer of abstraction that lets us interact with arbitrary storage mechanisms as if
they are conventional file systems. It makes S3 'look' like a conventional file system.

## Anonymous Reads

> A lot of data is available for global read, which does not require credentials or a profile.
> In this case, we set `anon=True` when plumbing the `fsspec` object:

Reviewer: ```suggestion
A lot of data is available for global read, which does not require credentials or a profile.
In this case, we set `anon=True` when plumbing the `fsspec` object. Here we demonstrate how to list the contents of an AWS S3 bucket with global read access:
```

```python
fs = fsspec.filesystem(
    's3',
    anon=True   # Does not require credentials
    )
# 'fs' is an object which provides methods access to the virtual filesystem
# 'ls' method == list ; list files in a virtual folder to test if we really have access

> fs.ls('s3://noaa-nwm-retrospective-2-1-zarr-pds/')

Reviewer: is the URL the same for data on the OSN pod?

        line = f.readline()
        print(line)
```

Output:

```text
b'<!DOCTYPE html>\r\n'
b'\r\n'
b'<!--\r\n'
b'Copyright 2014-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.\r\n'
b'\r\n'
```

However -- most of the time, we don't need such low level access.
A more general, and useful, mechanism is to '_map_' an S3 object to a _virtual file_ using `get_mapper()`.
That _map_ object can then be treated as if it is a local file -- the `fsspec` mapping mechanism handles
all of the translation to interact with S3.

```python
m = fs.get_mapper('s3://noaa-nwm-retrospective-2-1-zarr-pds/chrtout.zarr')
# 'm' looks like a file as far as Python is concerned.
# It is now a stand-in for those functions that require a file as argument.
g = zarr.convenience.open_consolidated(m) # read zarr metadata for named file.
print(g.tree())
```

Output:

```text
/
 ├── crs () |S1
 ├── elevation (2776738,) float32
 ├── feature_id (2776738,) int32
 ├── gage_id (2776738,) |S15
 ├── latitude (2776738,) float32
 ├── longitude (2776738,) float32
 ├── order (2776738,) int32
 ├── streamflow (367439, 2776738) int32
 ├── time (367439,) int64
 └── velocity (367439, 2776738) int32
```

> :::{warning}

Reviewer: when I view this as rendered markdown, these are still just showing up as text

Author: When you say "these" -- I think you are referring to `{warning}` ... if that's not right, then I need to understand more about your observation. 

The `warning` directive is not being parsed by the markdown renderer. That's correct -- because it isn't markdown.  This is a **rST** directive to be used when the notebook is assembled by jupyterbook. 

This may help: <https://github.com/hytest-org/hytest/blob/gt-118-JupyterBook-TOC/doc/JupyterBook_Admin.md> 

Some features in the fully-built jupterbook will not render in vanilla markdown.  The directives are constructed such that they will still render as text (just not very pretty, as you observed) in a markdown context, but will have full features when jupyterbook parses them.

Reviewer: yes - sorry for the ambiguity, I was referring to :::{warning} and :::{sidebar}

It would be nice if the files rendered correctly in markdown because it seems likely that I would just send someone a link to this markdown file on our Github to read....but maybe I will be convinced otherwise once we have the JupyterBook working. I would say, leave as is for now, and once we merge in the JupyterBook TOC, we can revisit if these headers not rendering on the .md file is a concern. I am trying to get through the PRs today if possible!

Author: Is there an advantage to sending someone a link to this markdown on github vs sending them a link to the fully-featured HTML that jupyterbook produces from it (also on github)?

Author: If we are not going to send people to the fully-featured, rendered jupyterbook, then why go to the trouble of producing it?  If vanilla markdown meets the mission of the project, we can save a lot of complexity by staying within its limitations and avoid jupyterbook altogether.

Reviewer: Hopefully we will always send links to the Jupyterbook, but I've never worked on a project that produced one before, so I'm just acknowledging my ignorance of any nuances that may come with it. I think we can assume we will use the JupyterBook moving forward, but I'm making a mental note to think about it a little more once we have it :)

We are deliberately demonstrating using commands that read **only** the zarr metadata from the
object storage -- **not the full data set**.

The above example is a very, _very_, **very** large dataset, which you don't want to load over the network to your desktop.
Execute full data read operations only if this notebook is being hosted and run out of the same AWS center where the data lives.

:::

:::{sidebar}

The good news about some of the larger science-oriented libraries (xarray, dask, pandas, zarr, etc), is that
they can automatically handle the `fsspec` operations for you **IF YOUR ACCESS IS ANONYMOUS**.

> This is a convenience, but is a special case for read-only data where `anon=True` and `requestor_pays=False`.

Reviewer: ```suggestion
This is a convenience, but is a special case for read-only data where `anon=True` and `requester_pays=False`.
```

Note that this is a feature of
[specific libraries](https://filesystem-spec.readthedocs.io/en/latest/#who-uses-fsspec),
and doesn't work everywhere.

Because it isn't universally available, and only applies to anonymous reads, examples in HyTEST will
always explicitly plumb `fsspec` 'longhand' using `get_mapper()`. You may see example code elsewhere
that takes the shortcut if it is available.

:::

## Credentialed Access

> For most data storage within the HyTEST workflows, access will not be anonymous.

Reviewer: I don't want to assume this will always be true!
```suggestion
For some data storage within the HyTEST workflows, access will not be anonymous.
```

Permissions are set by the owners of that data, using credentials assigned to an AWS 'profile'.

Profile credentials are usually stored outside of the Python program (typically in a master file in

> your ``HOME` folder on the compute/jupyter server).

Reviewer: ```suggestion
your `HOME` folder on the compute/jupyter server).
```

> Profile credentials are usually stored outside of the Python program (typically in a master file in
> your `HOME` folder on the compute/jupyter server).
> You need to have this set up beforehand, and is usually achieved by copying specific credential files into the right spot.
> A common mechanism to handle the profile configuration is with the `aws`
> [command line interface](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/configure/index.html).

Reviewer: ```suggestion
Profile credentials are usually stored outside of the Python program for security purposes (typically in a file in
your `HOME` folder on the compute/jupyter server).
You need to have this credential file set up before you can work with data in buckets requiring credentials. 
We will configure our OSN pod credentials with the `aws`
[command line interface](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/configure/index.html).
```

> To create a new AWS profile:

Reviewer: ```suggestion
To create a new AWS profile with the name `osn-renci`:
```

```sh
> aws configure --profile osn-renci
AWS Access Key ID :
AWS Secret Access Key :
Default region name : us-east-1
Default output format: json
```

> The first two prompts will ask for key/security information you were assigned for access to object storage.

Reviewer: How do we get these? Maybe this is something I should learn from Rich, but we might want to mention that the reader should be working with our team if they are writing to this bucket, and we will provide the credentials.

Reviewer: @rsignell-usgs - I need to make sure I have credentials with admin permissions on OSN before you disappear :) - not sure who else might already have these?

> The first two prompts will ask for key/security information you were assigned for access to object storage.

Reviewer: ```suggestion
The first two prompts will ask for key/security information you were assigned for access to object storage. Credentials for the OSN pod should have been provided by the HyTEST team if you need credentialed access to the OSN pod.
```

> Answer last questions with "us-east-1" (region) and "json" (format).

Reviewer: ```suggestion
The default region for the OSN pod should be "us-east-1" because this is the region in which the pod is physically located, and default output format should be "json".
```

> Note that this configuration is specific to the OSN 'pod' storage.

Reviewer: Can you add a little info above about what the OSN pod is and how it relates to S3, probably near the beginning of this doc so people understand that this is demoing how to use OSN and not S3? Especially since the intro makes it sound like these are instructions for S3 (I know the process is very similar, but maybe we need a little more framing). We might also want to rename the file!

Reviewer: I believe we want to say that HyTEST has procured an OSN storage pod (link) which is a ______ (something about a piece of hardware we own with object storage on a high speed network - @rsignell-usgs help me out with words here). It is similar to S3 object storage, but given that we own the hardware, we do not have to charge storage or egress fees. The OSN pod is compatible with most S3 API commands, but there are a few details that differ for reading/writing to the OSN pod, which this document will cover.

Reviewer: Would also love to link in any documentation for interfacing with the pod. @rsignell-usgs may have a good link for this too.

Author: @amsnyder -- I am not the right person to talk about OSN's architecture, funding, ownership, who pays what egress fees, etc. Mostly because I don't know any of that information. When I have questions about what OSN is and how it works, I'd go to @rsignell-usgs ... perhaps he can blurb the information you want here.

> Your profile name and region may be different.

Reviewer: ```suggestion
Your profile name and region may be different if you are setting up your credentials for an AWS S3 object storage bucket.
```

## Endpoints

> For storage operations, the AWS system needs the web address of the access point, or _endpoint_

Reviewer: ```suggestion
For storage operations, the S3 API needs the web address of the access point, or _endpoint_
```

where it should address filesystem operations. If your storage is completely within the Amazon
ecosystem, you will likely not need to specify an endpoint.
However, for 3rd-party storage (such as the OSN pod), you will need to explicitly declare the
endpoint when the filesystem is first referenced using `fsspec`.

```python
fs_write = fsspec.filesystem(
    's3',
    profile='osn-renci',  ## This is the profile name you configured above.
    client_kwargs={'endpoint_url': 'https://renc.osn.xsede.org'}
)
```

The enpoint URL will be given to you when your key ID and access key are assigned.

## Writing Data to S3

> With greater permissions, you may be able to do more destructive activities (overwriting, removing, etc).

Reviewer: greater permissions than what? What are default permissions for a new user? How would someone get these elevated permissions?

Author: greater permission than read-only/anonymous.

> With greater permissions, you may be able to do more destructive activities (overwriting, removing, etc).

Reviewer: ```suggestion
With adequate permissions, you may be able to do more destructive activities to objects in a bucket (overwriting, removing, etc).
```

Examples:

* `mkdir` -- makes a new directory / folder
* `mv` -- moves/renames a file or folder
* `rm` -- removes a file or folder

See the [API documentation](https://filesystem-spec.readthedocs.io/en/latest/api.html)
for the full details of available operations.

> ## Read-only access with "requestor pays"

Reviewer: ```suggestion
## Read-only access with "requester pays"
```

Some datasets, such as the first example above (anonymous reading), are offered with 'egress fees' paid by the data owner.
This means that all access is free to anybody.
The cost of the network bandwidth used to serve the data is covered by the data's host.
Not all "public" data is offered this way: it is still available to anybody who wants to read it, but

> the access fees must be paid by the reador (i.e. the 'requestor').

Reviewer: ```suggestion
the access fees must be paid by the reador (i.e. the 'requester').
```

> the access fees must be paid by the reador (i.e. the 'requester').

Reviewer: ```suggestion
the access fees must be paid by the reader (i.e. the 'requester').
```

> When you access a requestor-pays dataset, your profile identifies the account which will be

Reviewer: ```suggestion
When you access a requester-pays dataset, your profile identifies the account which will be
```

billed for access.  Open such datasets with an extra option to `fsspec`:

```python
fs = fsspec.filesystem(
    's3',
    profile='osn-renci',
    anon=False,

>     requestor_pays=True

Reviewer: does OSN pod use requestor? S3 uses requester
```suggestion
    requester_pays=True
```

[fsspec](https://filesystem-spec.readthedocs.io/en/latest/) module.
This is a layer of abstraction that lets us interact with arbitrary storage mechanisms as if
they are conventional file systems. It makes this object storage 'look' like a conventional file system.

## Anonymous Reads

A lot of data is available for global read, which does not require credentials or a profile.
In this case, we set `anon=True` when plumbing the `fsspec` object. Here we demonstrate how to list the contents of an AWS S3 bucket with global read access:

```python
fs = fsspec.filesystem(
    's3',
    anon=True   # Does not require credentials
    )
# 'fs' is an object which provides methods access to the virtual filesystem
# 'ls' method == list ; list files in a virtual folder to test if we really have access
fs.ls('s3://noaa-nwm-retrospective-2-1-zarr-pds/')
```

Output:

```text
['noaa-nwm-retrospective-2-1-zarr-pds/chrtout.zarr',
 'noaa-nwm-retrospective-2-1-zarr-pds/gwout.zarr',
 'noaa-nwm-retrospective-2-1-zarr-pds/lakeout.zarr',
 'noaa-nwm-retrospective-2-1-zarr-pds/ldasout.zarr',
 'noaa-nwm-retrospective-2-1-zarr-pds/precip.zarr',
 'noaa-nwm-retrospective-2-1-zarr-pds/rtout.zarr']
```

> 

Reviewer: I decided to call this `fs_osn` to not confuse it with the `fs` set up for S3 above that is used in some of the below commands. Open to other ways of framing this though!

b'\r\n'
b'<!--\r\n'
b'Copyright 2014-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.\r\n'
b'\r\n'
```

However -- most of the time, we don't need such low level access.
A more general, and useful, mechanism is to '_map_' an S3 object to a _virtual file_ using `get_mapper()`.
That _map_ object can then be treated as if it is a local file -- the `fsspec` mapping mechanism handles
all of the translation to interact with S3.

```python
m = fs.get_mapper('s3://noaa-nwm-retrospective-2-1-zarr-pds/chrtout.zarr')
# 'm' looks like a file as far as Python is concerned.
# It is now a stand-in for those functions that require a file as argument.
g = zarr.convenience.open_consolidated(m) # read zarr metadata for named file.
print(g.tree())
```

Output:

```text
/
 ├── crs () |S1
 ├── elevation (2776738,) float32
 ├── feature_id (2776738,) int32
 ├── gage_id (2776738,) |S15
 ├── latitude (2776738,) float32
 ├── longitude (2776738,) float32
 ├── order (2776738,) int32
 ├── streamflow (367439, 2776738) int32
 ├── time (367439,) int64
 └── velocity (367439, 2776738) int32
```

:::{warning}

We are deliberately demonstrating using commands that read **only** the zarr metadata from the

> object storage -- **not the full data set**.

Reviewer: updating spelling for consistency
```suggestion
object storage -- **not the full dataset**.
```

## Endpoints

For storage operations, the S3 API needs the web address of the access point, or _endpoint_
where it should address filesystem operations. If your storage is completely within the Amazon
ecosystem, you will likely not need to specify an endpoint.
However, for 3rd-party storage (such as the OSN pod), you will need to explicitly declare the
endpoint when the filesystem is first referenced using `fsspec`.

```python
fs_write = fsspec.filesystem(
    's3',
    profile='osn-renci',  ## This is the profile name you configured above.
    client_kwargs={'endpoint_url': 'https://renc.osn.xsede.org'}
)
```

The enpoint URL will be given to you when your key ID and access key are assigned.

## Writing Data to S3

With adequate permissions, you may be able to do more destructive activities to objects in a bucket (overwriting, removing, etc).
Examples:

* `mkdir` -- makes a new directory / folder
* `mv` -- moves/renames a file or folder
* `rm` -- removes a file or folder

From within your Python program, writes to S3 storage can be achieved a few different ways.
Often, the most convenient is to use a `mapper` to connect a file-like python object to

> the S3 object storage location:

Reviewer: ```suggestion
the object storage location:
```

object storage -- **not the full data set**.

The above example is a very, _very_, **very** large dataset, which you don't want to load over the network to your desktop.
Execute full data read operations only if this notebook is being hosted and run out of the same AWS center where the data lives.

:::

:::{sidebar}

The good news about some of the larger science-oriented libraries (xarray, dask, pandas, zarr, etc), is that
they can automatically handle the `fsspec` operations for you **IF YOUR ACCESS IS ANONYMOUS**.
This is a convenience, but is a special case for read-only data where `anon=True` and `requester_pays=False`.

Note that this is a feature of
[specific libraries](https://filesystem-spec.readthedocs.io/en/latest/#who-uses-fsspec),
and doesn't work everywhere.

Because it isn't universally available, and only applies to anonymous reads, examples in HyTEST will
always explicitly plumb `fsspec` 'longhand' using `get_mapper()`. You may see example code elsewhere
that takes the shortcut if it is available.

:::

## Credentialed Access

For some data storage within the HyTEST workflows, access will not be anonymous.

> Permissions are set by the owners of that data, using credentials assigned to an AWS 'profile'.

Reviewer: ```suggestion
Permissions are set by the owners of that data, and the rules governing your ability to read from or write to certain locations may be defined with a set of credentials assigned to an AWS 'profile'.
```

Profile credentials are usually stored outside of the Python program (typically in a master file in
your `HOME` folder on the compute/jupyter server).
You need to have this set up beforehand, and is usually achieved by copying specific credential files into the right spot.
A common mechanism to handle the profile configuration is with the `aws`
[command line interface](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/configure/index.html).

To create a new AWS profile:

```sh

> > aws configure --profile osn-renci

Reviewer: ```suggestion
> aws configure --profile osn-scratch
```

## Endpoints

For storage operations, the AWS system needs the web address of the access point, or _endpoint_
where it should address filesystem operations. If your storage is completely within the Amazon
ecosystem, you will likely not need to specify an endpoint.
However, for 3rd-party storage (such as the OSN pod), you will need to explicitly declare the
endpoint when the filesystem is first referenced using `fsspec`.

```python
fs_write = fsspec.filesystem(
    's3',
    profile='osn-renci',  ## This is the profile name you configured above.
    client_kwargs={'endpoint_url': 'https://renc.osn.xsede.org'}
)
```

The enpoint URL will be given to you when your key ID and access key are assigned.

## Writing Data to S3

With greater permissions, you may be able to do more destructive activities (overwriting, removing, etc).
Examples:

* `mkdir` -- makes a new directory / folder
* `mv` -- moves/renames a file or folder
* `rm` -- removes a file or folder

From within your Python program, writes to S3 storage can be achieved a few different ways.
Often, the most convenient is to use a `mapper` to connect a file-like python object to
the S3 object storage location:

```python

> fname='s3://rsignellbucket2/testing/outfile.zarr'
> outfile=fs_write.get_mapper(fname)
> xarray_dataset.to_zarr(outfile)

Reviewer: Updating the path to use our scratch bucket we set up. And I needed these two extra arguments in the `to_zarr()` command to make this work. 
```suggestion
fname='usgs-scratch/testing/outfile.zarr'
outfile=fs_write.get_mapper(fname)
xarray_dataset.to_zarr(outfile, mode='w', consolidated=True)
```

```

The `outfile` variable can be used most anywhere that a file-like object is needed for
writing.

See the [API documentation](https://filesystem-spec.readthedocs.io/en/latest/api.html)
for the full details of available operations.

## Read-only access with "requester pays"

Some datasets, such as the first example above (anonymous reading), are offered with 'egress fees' paid by the data owner.
This means that all access is free to anybody.
The cost of the network bandwidth used to serve the data is covered by the data's host.
Not all "public" data is offered this way: it is still available to anybody who wants to read it, but
the access fees must be paid by the reador (i.e. the 'requester').

When you access a requester-pays dataset, your profile identifies the account which will be
billed for access.  Open such datasets with an extra option to `fsspec`:

```python
fs = fsspec.filesystem(
    's3',
    profile='osn-renci',
    anon=False,
    requester_pays=True
)

> fs.ls('s3://esip-qhub/noaa/nwm/')

Reviewer: Suggesting we use an OSN pod endpoint here since the fsspec fs was set up for that above.
```suggestion
fs.ls('usgs-scratch/')
```
